pkgs/applications/virtualization/singularity/generic.nix

```diff
@@ -0,0 +1,199 @@
+{ lib
+, buildGoModule
+, cryptsetup
+, makeWrapper
+, util-linux
+, which
+, bash
+, coreutils
+, go
+, gpgme
+, libuuid
+  # This is for nvidia-container-cli
+, nvidia-docker
+, openssl
+, squashfsTools
+  # Configurations
+, pname
+, version
+, src
+, projectName # "apptainer" or "singularity"
+  # These cannot be correctly override with `overrideAttrs`
+, vendorSha256 ? null
+, deleteVendor ? false
+, proxyVendor ? false
+, doCheck ? true
+, enableNvidiaContainerCli ? true
+  # This is required to run containers without suid
+, withoutSuid ? true
+, starterSuidPath ? null
+  # Remove the symlink to `singularity` when projectName != "singularity"
+, removeCompat ? true
+, extraDescription ? ""
+, extraMeta ? { }
+}:
+
+buildGoModule {
+  inherit pname version src;
+
+  # All three projects provide vendored source tarball
+  # so we set vendorSha256 = null;
+  # One who use unvendored source can override this
+  # with `override`
+  inherit vendorSha256 deleteVendor proxyVendor;
+
+  # apptainer and singularity-ce both want to reference Go
+  # to compile the extensions when building container images
+  allowGoReference = true;
+
+  passthru = {
+    inherit
+      projectName
+      removeCompat
+      withoutSuid
+      ;
+  };
+
+  buildInputs = [
+    gpgme
+    libuuid
+    openssl
+  ];
+
+  nativeBuildInputs = [
+    cryptsetup
+    makeWrapper
+    util-linux
+    which
+  ];
+
+  postPatch = ''
+    substituteInPlace internal/pkg/build/files/copy.go \
+      --replace /bin/cp ${coreutils}/bin/cp
+  '';
+
+  configureScript = "./mconfig";
+
+  configureFlags = [
+    "--localstatedir=/var"
+  ]
+  ++ lib.optional withoutSuid "--without-suid";
+
+  preConfigure = ''
+    if ! [ -e .git ] || ! [ -e VERSION ]; then
+      echo "${version}" > VERSION
+    fi
+  '';
+
+  postConfigure = ''
+    patchShebangs .
```

Comment:
We should be more specific about the files that need checking because otherwise we try to patch every file which can be a bit slow.